File: src/array/python/49.py
from collections import defaultdict


class Solution:
    def groupAnagrams(self, strs):
        anagram_map = defaultdict(list)

        for word in strs:
            sorted_word = ''.join(sorted(word))
            anagram_map[sorted_word].append(word)

        return list(anagram_map.values())


s = Solution()

# order does not matter, so errors are ok
assert s.groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]) == [["bat"], ["nat", "tan"], ["ate", "eat", "tea"]]
assert s.groupAnagrams([""]) == [[""]]
assert s.groupAnagrams(["a"]) == [["a"]]
assert s.groupAnagrams(["ddddddddddg", "dgggggggggg"]) == [["dgggggggggg"], ["ddddddddddg"]]

File: src/array/cpp/49.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> result;
        unordered_map<string, vector<string>> anagrams;
        for (const string& s: strs) {
            string sorted = s;
            sort(sorted.begin(), sorted.end());
            anagrams[sorted].push_back(s);
        }
        for (const auto& p: anagrams) {
            result.push_back(p.second);
        }
        return result;
    };
};


int main() {
    Solution s;

    vector<string> test1 = {"eat", "tea", "tan", "ate", "nat", "bat"};
    vector<vector<string>> result1 = s.groupAnagrams(test1);
    assert(result1.size() == 3 && "Test 1 Failed: Size mismatch");

    vector<string> test2 = {""};
    vector<vector<string>> result2 = s.groupAnagrams(test2);
    assert(result2.size() == 1 && result2[0].size() == 1 && result2[0][0] == "" && "Test 2 Failed");

    vector<string> test3 = {"a"};
    vector<vector<string>> result3 = s.groupAnagrams(test3);
    assert(result3.size() == 1 && result3[0].size() == 1 && result3[0][0] == "a" && "Test 3 Failed");

    cout << "All tests passed!" << endl;
}